### PR TITLE
Firefox with extensions and global config

### PR DESCRIPTION
**Pull request by Qubasa**
This pull request adds two main features.
1. To be able to configure Firefox extensions through nix packages
2. Added many configuration options like

```
  , extraPrefs ? ""
  , extraExtensions ? [ ]
  , noNewProfileOnFFUpdate ? false
  , allowNonSigned ? false
  , disablePocket ? false
  , disableTelemetry ? true
  , disableDrmPlugin ? false
  , showPunycodeUrls ? true
  , enableUserchromeCSS ? false
  , disableFirefoxStudies ? true
  , disableFirefoxSync ? false
  , disableFirefoxUpdatePage ? true
  , useSystemCertificates ? true
  , dontCheckDefaultBrowser ? false
  # For more information about anti tracking (german website)
  # vist https://wiki.kairaven.de/open/app/firefox
  , activateAntiTracking ? true
  , disableFeedbackCommands ? true
  , disableDNSOverHTTPS ? true
  , disableGoogleSafebrowsing ? true
  , clearDataOnShutdown ? false
  , homepage ? "about:home"
  , enableDarkDevTools ? false
  # For more information about policies visit
  # https://github.com/mozilla/policy-templates#enterprisepoliciesenabled
  , extraPolicies ? {}
```

I used and added features to the code from https://github.com/NixOS/nixpkgs/pull/57554 as a reference as the pull request is sadly on hold.

An example for a Firefox extension nix package would be:
https://github.com/Luis-Hebendanz/nix-configs/blob/master/own-pkgs/ublock-origin/default.nix
or if the extension does not provide an extension id it would be like this:
https://github.com/Luis-Hebendanz/nix-configs/blob/master/own-pkgs/canvas-fingerprint-defender/default.nix

An example on how to call the wrapper is located here:
https://github.com/Luis-Hebendanz/nix-configs/blob/master/firefox.nix

I am sure that some things can be improved.  Sadly I have very limited time but many people want this so I am putting this up. If you have improvement ideas please add them directly as a commit because I do not have the time to implement them myself ^^

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ x  ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Lassulus

### CODE REVIEW

**Comment by jtojnar** [comment]
It looks like you did not push the commits in question. Please fix the branch locally force push to update it here.

Also I am not sure how well will adding all the config options as top-level arguments scale – maybe grouping it under common attrset will be nicer.

**Comment by 7c6f434c** [comment]
>  # For more information about anti tracking
>  # vist https://wiki.kairaven.de/open/app/firefox

Maybe mention that the page is in German?

>  , activateAntiTracking ? true

What is the promise here? Enable the Enterprise policy? Policy plus some extra about:config settings? Will the set of extras be updated?

I think a nice first step would be to have policies and config entries as two attrset arguments, with commented defaults and `deepMerge` of the arguments on top of the defaults. And use exactly the upstream names/paths.

>  , enableDarkDevTools ? true

This is just colour scheme application? Why set it to not-upstream-default?

Maybe also add a default-off option for honouring userChrome.css?

**Comment by andir** [comment]
The Firefox wrapper is reaching a point where we should probably have tests for it. Otherwise this looks great but I do not think anyone is happily maintaining this without some supportive tests for (each of?) the features.

Also did you solve the uninstall of firefox extensions? Previous attempts always fell short on that and I do not feel like adding support for adding but not removing plugins.

**Comment by Luis-Hebendanz** [comment]
> # For more information about anti tracking # vist https://wiki.kairaven.de/open/app/firefox
> Maybe mention that the page is in German?
> , activateAntiTracking ? true
> What is the promise here? Enable the Enterprise policy? Policy plus some extra about:config settings? Will the set of extras be updated? I think a nice first step would be to have policies and config entries as two attrset arguments, with commented defaults and `deepMerge` of the arguments on top of the defaults. And use exactly the upstream names/paths.
> , enableDarkDevTools ? true
> This is just colour scheme application? Why set it to not-upstream-default? Maybe also add a default-off option for honouring userChrome.css?

I changed the enableDarkDevTools to false and made a note that the link is a german website.
I also added the options "enableUserchromeCSS".

This is already the case the options `extraPrefs` and `extraPolicies` do exactly that. The other options are convenience features that I think are important to make visible and abstract away. Because otherwise nobody knows these options are available or how exactly to set them ( as often times you have to set multiple settings in userPref.json and in the enterprise Policy to have the desired effect)

**Comment by Luis-Hebendanz** [comment]
> The Firefox wrapper is reaching a point where we should probably have tests for it. Otherwise this looks great but I do not think anyone is happily maintaining this without some supportive tests for (each of?) the features.
> 
> Also did you solve the uninstall of firefox extensions? Previous attempts always fell short on that and I do not feel like adding support for adding but not removing plugins.
I agree having tests for it is probably a good idea. Sadly I don't have the time and resources to do that but maybe someone else?
I also thought about nicely asking Mozilla  to update the userPref.json attributes, as they do provide the wrapper and firefox package for nixos and have the most in depth knowledge about it (of course)

**Comment by jtojnar** [comment]
> This is already the case. The options `extraPrefs` and `extraPolicies` do exactly that. The other options are convenience features that I think are important to make visible and abstract away. Because otherwise nobody knows these options are available or how exactly to set them ( as often times you have to set multiple settings in userPref.json and in the enterprise Policy to have the desired effect)

I am still not convinced the package should expose convenience features. Could we have the low-level options and let the convenience be handled by external helpers? Something like this overlay:

```nix
(self: super: {
  firefoxCustomization.enableDarkDevTools = config: super.lib.recursiveUpdate config {
    lockPrefs = {
      "devtools.theme" = "dark";
    };
  };

  firefoxCustomization.disableManualExtensions = config: super.lib.recursiveUpdate config {
    policies = {
        ExtensionSettings = {
          "*" = {
            blocked_install_message = "You can't have manual extension mixed with nix extensions";
            installation_mode = "blocked";
          };
        };
    };
    lockPrefs = {
      "extensions.getAddons.showPane" = false;
      "extensions.htmlaboutaddons.recommendations.enabled" = false;
      "app.update.auto" = false;
    };
  };

  myFirefox = super.firefox.override {
    config = super.lib.pipe super.firefox.defaultConfig [
      self.firefoxCustomization.enableDarkDevTools
      self.firefoxCustomization.disableManualExtensions
    ];
  };
})
```

**Comment by Luis-Hebendanz** [comment]
> > This is already the case. The options `extraPrefs` and `extraPolicies` do exactly that. The other options are convenience features that I think are important to make visible and abstract away. Because otherwise nobody knows these options are available or how exactly to set them ( as often times you have to set multiple settings in userPref.json and in the enterprise Policy to have the desired effect)
> 
> I am still not convinced the package should expose convenience features. Could we have the low-level options and let the convenience be handled by external helpers? Something like this overlay:
> 
> ```nix
> (self: super: {
>   firefoxCustomization.enableDarkDevTools = config: super.lib.recursiveUpdate config {
>     lockPrefs = {
>       "devtools.theme" = "dark";
>     };
>   };
> 
>   firefoxCustomization.disableManualExtensions = config: super.lib.recursiveUpdate config {
>     policies = {
>         ExtensionSettings = {
>           "*" = {
>             blocked_install_message = "You can't have manual extension mixed with nix extensions";
>             installation_mode = "blocked";
>           };
>         };
>     };
>     lockPrefs = {
>       "extensions.getAddons.showPane" = false;
>       "extensions.htmlaboutaddons.recommendations.enabled" = false;
>       "app.update.auto" = false;
>     };
>   };
> 
>   myFirefox = super.firefox.override {
>     config = super.lib.pipe super.firefox.defaultConfig [
>       self.firefoxCustomization.enableDarkDevTools
>       self.firefoxCustomization.disableManualExtensions
>     ];
>   };
> })
> ```

Ohh this is nice. Yeah seems like a good idea

**Comment by Luis-Hebendanz** [comment]
Sorry somehow I cherry picked the wrong commit version. I added now everything that was missing.

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,8 @@ result-*
 /pkgs/development/libraries/qt-5/*/tmp/
 /pkgs/desktops/kde-5/*/tmp/
 /pkgs/development/mobile/androidenv/xml/*
+.envrc
+.direnv
+shell.nix
+.history
+pkgs/applications/search/deezer-downloader/
```
Comment (7c6f434c) [inline .gitignore]: I don't have anything against expanding gitignore, but another-package-specific entry is probably too unrelted

File: pkgs/applications/networking/browsers/firefox/wrapper.nix [hunk 2]
```diff
@@ -83,6 +113,245 @@ let
...
+        else if a.signed == false && !allowNonSigned then
+          throw "Disable signature checking in firefox if you want ${a.pname} addon"
+        else  a
+      ) extraExtensions;
+
+      policiesJson = builtins.toFile "policies.json"
+        (builtins.toJSON enterprisePolicies);
+
+      mozillaCfg = builtins.toFile "mozilla.cfg" ''
+// First line must be a comment
```
Comment (7c6f434c) [inline pkgs/applications/networking/browsers/firefox/wrapper.nix]: I think `''` strips indentation if it is equal throught the string, so no need to have zero indentation here

File: pkgs/applications/networking/browsers/firefox/wrapper.nix [hunk 3]
```diff
@@ -83,6 +113,245 @@ let
...
+
+        // Remove default top sites
+        lockPref("browser.newtabpage.pinned", "");
+        lockPref("browser.newtabpage.activity-stream.default.sites", "");
```
Comment (7c6f434c) [inline pkgs/applications/networking/browsers/firefox/wrapper.nix]: If we generate the policy unconditionally, shouldn't disbling any functionlity be conditional on some flag?

**Comment by bjornfor** [comment]
Oh, sad to see this closed. Didn't look at the implementation, but the feature is wanted (by me at least).

**Comment by Luis-Hebendanz** [comment]
I made an overhauled version of it: https://github.com/NixOS/nixpkgs/pull/91724

**Pull request closed without merging**